framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4240,6 +4291,203 @@ private void makeGroundTargetType(
         addDefaultAnnotations(functionalType);
     }
 
+    /**
+     * Return true if {@code type} should be captured.
+     *
+     * <ul>
+     *   <li>{@code type} and {@code typeMirror} must both be declared types.
+     *   <li>{@code type} must not be a raw type, must not have an uninferred type argument, and
+     *       must have a wildcard as a type argument.
+     *   <li>{@code typeMirror} must have a captured type as a type argument.
+     * </ul>
+     *
+     * @param type annotated type to capture
+     * @param typeMirror underlying Java type
+     * @return true if {@code type} should be captured
+     */
+    private boolean shouldCapture(AnnotatedTypeMirror type, TypeMirror typeMirror) {
+        if (!(type.getKind() == TypeKind.DECLARED && typeMirror.getKind() == TypeKind.DECLARED)) {
+            return false;
+        }
+
+        DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+        AnnotatedDeclaredType typeToCapture = (AnnotatedDeclaredType) type;
+        if (typeToCapture.wasRaw() || typeToCapture.containsUninferredTypeArguments()) {
+            return false;
+        }
+
+        if (typeToCapture.getTypeArguments().isEmpty()
+                || (capturedTypeMirror.getTypeArguments().size()
```

Comment:
Why is this size comparison necessary as part of the test?  I think that if it is true, there is some corruption.  Should you change this into an assertion?